backend/src/Edna/Web/Types.hs

```diff
@@ -17,4 +43,180 @@ data ExperimentalMeasurement = ExperimentalMeasurement
   , emOutlier :: Bool
   } deriving stock (Generic, Show, Eq)
 
-deriveJSON ednaAesonWebOptions ''ExperimentalMeasurement
+----------------
+-- General types
+----------------
+
+-- | A simple wrapper over 'Word'. At the data layer, we identify all entities
+-- with numeric IDs and this data type corresponds to such an ID.
+-- It has a phantom parameter type which the type of the object
+-- identified by this ID.
+newtype SqlId t = SqlId
+  { unSqlId :: Word
+  } deriving stock (Generic, Show)
+    deriving newtype (FromHttpApiData, FromJSON, ToJSON, ToSchema)
+
+-- | This data type is useful when you want to return something with its ID.
+data WithId t = WithId
+  { wiId :: SqlId t
+  , wItem :: t
+  } deriving stock (Generic, Show)
+
+-- | This data type is used when you want to return something with its ID and
+-- some additional data that was not submitted by end users, but is maintained
+-- by the application.
+data WithExtra t e = WithExtra
+  { weId :: SqlId t
+  , weItem :: t
+  , weExtra :: e
+  } deriving stock (Generic, Show)
+
+-- | A stub to specify the sorting order, most likely will be replaced with
+-- @servant-util@.
+data StubSortBy =
+    SortByName
+  | SortBySomething
+
+instance FromHttpApiData StubSortBy where
+  parseQueryParam = \case
+    "name" -> pure SortByName
+    "something" -> pure SortBySomething
+    x -> Left $ "unknown sorting order: " <> x
+
+----------------
+-- Entities
+----------------
+
+-- | Input data submitted along with uploaded file.
+data FileUploadInput = FileUploadInput
+  { fuiProject :: SqlId Project
+  -- ^ ID of the project the file belongs to.
+  , fuiTestMethodology :: SqlId TestMethodology
+  -- ^ ID of the test methodology used throughout the file.
+  } deriving stock (Generic, Show)
+
+-- | Summary of an experiment data file.
+newtype FileSummary = FileSummary
+  { unFileSummary :: [FileSummaryItem]
+  } deriving stock (Generic, Show)
+
+-- | One element in 'FileSummary'. Corresponds to one target from the file.
+-- Contains all compounds that interact with the target in the file.
+-- Also contains information whether this target is new or already known.
+data FileSummaryItem = FileSummaryItem
+  { fsiTarget :: SqlId Target
+  -- ID of a target from the file.
+  , fsiCompounds :: [SqlId Compound]
```

Comment:
@pva701 this is what I return from `parseFile`. The problem is that when the file is parsed, it's not saved to DB and some compounds and targets can be new. So they have no ID. So I can't return IDs in this case. Is it ok if I return names in both `fsiTarget` and `fsiCompounds`? Is it ok if I do the same in `uploadFile` or do you want `FileSummary` to be different there? Or maybe we don't need `FileSummary` in the output of `uploadFile` at all.

Reply:
yeah, the reasonable point. I think for now it's ok to return names from both endpoints. But I am not sure it's 100% final, as we are going to implement View button after file was parsed, maybe it will be better to return ID for the second endpoint to make it possible

Reply:
Ok, then maybe I'll return different types because AFAIU the current `FileSummaryItem` is good enough for `upload`, so I can just add a new type instead of _replacing_ this type.